src/rust/catnap/win/overlapped.rs

```diff
@@ -0,0 +1,774 @@
+                    );
+                }
+
+                match status {
+                    Ok(()) => {
+                        // Spurious wake-up.
+                        continue;
+                    },
+
+                    Err(err) => {
+                        if err.errno == libc::ECANCELED {
+                            if let Err(cancel_err) = cancel(completion.get_state_ref(), overlapped) {
+                                warn!("cancellation failed: {}", cancel_err);
+                            } else {
+                                completion.set_yielder(None);
+                                return Err(err);
+                            }
+                        }
+
+                        completion.abandon();
+                        return Err(err);
+                    },
+                }
+            },
+
+            // Operation failed to start.
+            Err(err) => Err(err),
+        }
+    }
+
+    /// Same as `do_io_with`, but does not use an intermediate state value.
+    pub async unsafe fn do_io<F1, F2, F3, R>(
+        &mut self,
+        yielder: &Yielder,
+        start: F1,
+        cancel: F2,
+        finish: F3,
+    ) -> Result<R, Fail>
+    where
+        F1: FnOnce(*mut OVERLAPPED) -> Result<(), Fail>,
+        F2: FnOnce(*mut OVERLAPPED) -> Result<(), Fail>,
+        F3: FnOnce(OverlappedResult) -> Result<R, Fail>,
+    {
+        self.do_io_with(
+            (),
+            yielder,
+            |_, overlapped: *mut OVERLAPPED| -> Result<(), Fail> { start(overlapped) },
+            |_, overlapped: *mut OVERLAPPED| -> Result<(), Fail> { cancel(overlapped) },
+            |_, result: OverlappedResult| -> Result<R, Fail> { finish(result) },
+        )
+        .await
+    }
+
+    /// Process a single overlapped entry.
+    #[allow(unused)]
+    fn process_overlapped(&mut self, entry: &OVERLAPPED_ENTRY) {
+        if let Some(overlapped) = std::ptr::NonNull::new(entry.lpOverlapped) {
+            // Safety: this is valid as long as the caller follows the contract: all queued OVERLAPPED instances are
+            // generated by `IoCompletionPort` API.
+            let overlapped: Pin<&mut OverlappedCompletion> = OverlappedCompletion::unmarshal(overlapped);
+
+            // Safety: the OVERLAPPED does not need to be pinned after being dequeued from the completion port.
+            let overlapped: &mut OverlappedCompletion = unsafe { overlapped.get_unchecked_mut() };
+            if let Some(mut yielder_handle) = overlapped.yielder_handle.take() {
+                debug_assert!(entry.dwNumberOfBytesTransferred as usize == overlapped.overlapped.InternalHigh);
+                overlapped.completion_key = entry.lpCompletionKey;
+                yielder_handle.wake_with(Ok(()));
+            } else {
+                // This can happen due to a failed cancellation or any other error on the do_overlapped path.
+                trace!("I/O dropped");
```

Comment:
Can some identifier be printed with this message? Or would the log around this automatically clarify?